### PR TITLE
[ISSUE #340]Add http trace http point

### PR DESCRIPTION
**Pull request by Roc-00**
Fixes ISSUE#<340>.

### Motivation

- new feature
- Integrate With OpenTelemetry's trace
- add http trace http point

### Modifications

Add class 'OpenTelemetryTraceFactory' which allow eventmeshHTTPServer to get tracer from it.
Rewrite the class 'loggingSpanExporter' of open telemetry in eventmesh,because it will casue garbled code.What I had done is change the logger into slf4j's logger.

### Documentation

- Does this pull request introduce a new feature? yes
- If yes, how is the feature documented? in markdown file.

### CODE REVIEW

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/trace/OpenTelemetryTraceFactory.java [hunk 1]
```diff
@@ -41,15 +44,18 @@
...
 
+    // Name of the service(using the instrumentationName)
+    private String SERVICE_NAME;
+
     public OpenTelemetryTraceFactory(CommonConfiguration configuration){
-        //different spanExporter
         try {
             String exporterName = configuration.eventMeshTraceExporterType;
-            spanExporter = (SpanExporter) Class.forName("io.opentelemetry.exporter."+exporterName+"."+exporterName+"SpanExporter").newInstance();
+            //todo different spanExporter
```
Comment (Roc-00) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/trace/OpenTelemetryTraceFactory.java:57]: I have no idea how to set different exporter.What should I do?

**Comment by xwm1992** [comment]
image
CommonConfigurationTest need to be fixed

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/exporter/logExporter.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * Flushes the data.
+     * (i guess it is not necessary for slf4j's log)
+     * @return the result of the operation
+     */
+    @Override
+    public CompletableResultCode flush() {
+        CompletableResultCode resultCode = new CompletableResultCode();
+//        for (Handler handler : logger.getHandlers()) {
```
Comment (xwm1992) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/exporter/logExporter.java:71]: remove the unused code

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/trace/OpenTelemetryTraceFactory.java (hunk 1 above)
Comment (xwm1992) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/trace/OpenTelemetryTraceFactory.java:57]: Different exporter may implement the `SpanExporter` and you can use the reflection to get different exporter instance

File: tool/license/allowed-licenses.txt [hunk 3]
```diff
@@ -564,6 +564,36 @@
...
+            "moduleVersion": "2.16.3",
+            "moduleName": "io.zipkin.reporter2:zipkin-sender-okhttp3"
+        },
+        {
+            "moduleLicense": "The Apache Software License, Version 2.0",
+            "moduleVersion": "2.23.2",
+            "moduleName": "io.zipkin.zipkin2:zipkin"
+        },
+        {
+            "moduleLicense": "Apache 2.0",
```
Comment (xwm1992) [inline tool/license/allowed-licenses.txt:589]: Here not correct, `The Apache Software License, Version 2.0` or  `The Apache License, Version 2.0`

File: tool/license/allowed-licenses.txt [hunk 4]
```diff
@@ -564,6 +564,36 @@
...
+            "moduleVersion": "3.14.9",
+            "moduleName": "com.squareup.okhttp3:okhttp"
+        },
+        {
+            "moduleLicense": "Apache 2.0",
```
Comment (xwm1992) [inline tool/license/allowed-licenses.txt:594]: same as above

**Review by xwm1992 — CHANGES_REQUESTED** [review]

File: tool/license/allowed-licenses.txt (hunk 3 above)
Comment (ruanwenjun) [reply tool/license/allowed-licenses.txt:589]: Yes, there might some problem with license check plugin, the one license may have different name, and we need to normalize them, we may need to add a LicenseBundleNormalizer  filter.
https://github.com/jk1/Gradle-License-Report

I will test this.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/exporter/logExporter.java (hunk 2 above)
Comment (Roc-00) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/exporter/logExporter.java]: ok

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/exporter/zipkinExporter.java [hunk 5]
```diff
@@ -17,8 +17,9 @@
 package org.apache.eventmesh.runtime.exporter;
 
 import io.opentelemetry.exporter.zipkin.ZipkinSpanExporter;
+import org.apache.eventmesh.common.config.CommonConfiguration;
 
-public class zipkinExporter {
+public class zipkinExporter implements EventMeshExporter {
```
Comment (xwm1992) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/exporter/zipkinExporter.java:22]: `zipkinExporter` isn't right, change to `ZipkinExporter`

File: eventmesh-runtime/conf/eventmesh.properties [hunk 6]
```diff
@@ -74,7 +74,7 @@ eventMesh.server.registry.enabled=false
 eventMesh.metrics.prometheus.port=19090
 
 #trace exporter
-eventmesh.trace.exporter.type=log
+eventmesh.trace.exporter.type=zipkin
```
Comment (xwm1992) [inline eventmesh-runtime/conf/eventmesh.properties:77]: here may be Zipkin, because of your class name

**Review by xwm1992 — CHANGES_REQUESTED** [review]

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/exporter/zipkinExporter.java (hunk 5 above)
Comment (Roc-00) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/exporter/zipkinExporter.java]: ok

File: eventmesh-runtime/conf/eventmesh.properties (hunk 6 above)
Comment (Roc-00) [reply eventmesh-runtime/conf/eventmesh.properties]: changed

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java [hunk 7]
```diff
@@ -132,13 +142,26 @@ public void sendError(ChannelHandlerContext ctx,
                 "; charset=" + EventMeshConstants.DEFAULT_CHARSET);
         response.headers().add(HttpHeaderNames.CONTENT_LENGTH, response.content().readableBytes());
         response.headers().add(HttpHeaderNames.CONNECTION, HttpHeaderValues.KEEP_ALIVE);
-        // todo server span end with error, record status, we should get channel here to get span in channel's context in async call..
+        //todo server span end with error, record status, we should get channel here to get span in channel's context in async call..
```
Comment (tydhot) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java:145]: You may need to add a switch to determine whether you want to enter here

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java [hunk 8]
```diff
@@ -132,13 +142,26 @@ public void sendError(ChannelHandlerContext ctx,
...
+            span.setStatus(StatusCode.ERROR);//set this span's status to ERROR
+            span.end();// closing the scope does not end the span, this has to be done manually
+        }
         ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
     }
 
     public void sendResponse(ChannelHandlerContext ctx,
                              DefaultFullHttpResponse response) {
-        // todo end server span, we should get channel here to get span in channel's context in async call.
+        //todo end server span, we should get channel here to get span in channel's context in async call.
```
Comment (tydhot) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java:157]: You may need to add a switch to determine whether you want to enter here

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java [hunk 9]
```diff
@@ -132,13 +142,26 @@ public void sendError(ChannelHandlerContext ctx,
...
+        Context context = ctx.channel().attr(AttributeKeys.SERVER_CONTEXT).get();
+        Span span = context.get(SpanKey.SERVER_KEY);
+        try (Scope ignored = context.makeCurrent()) {
+            span.addEvent("Send Response");
```
Comment (tydhot) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java:161]: no need to add event about response

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java [hunk 10]
```diff
@@ -256,7 +294,18 @@ protected void channelRead0(ChannelHandlerContext ctx, HttpRequest httpRequest)
                 requestCommand.setHttpMethod(httpRequest.method().name());
                 requestCommand.setHttpVersion(httpRequest.protocolVersion().protocolName());
                 requestCommand.setRequestCode(requestCode);
-                // todo record command method, version and requestCode in span.
+
+                //todo record command opaque in span.
+                span = tracer.spanBuilder("HTTP "+requestCommand.httpMethod).setParent(context).setSpanKind(SpanKind.SERVER).startSpan();
+                span.addEvent("Span Start");
```
Comment (tydhot) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java:300]: avoid span event here

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java [hunk 11]
```diff
@@ -256,7 +294,18 @@ protected void channelRead0(ChannelHandlerContext ctx, HttpRequest httpRequest)
...
+                //attach the span to the server context
+                context = context.with(SpanKey.SERVER_KEY,span);
+                //put the context in channel
+                ctx.channel().attr(AttributeKeys.SERVER_CONTEXT).set(context);
+                //todo record command method, version and requestCode in span.
+                span.setAttribute("HttpMethod",httpRequest.method().name());
```
Comment (tydhot) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java:306]: ues otel key instaed : https://github.com/open-telemetry/opentelemetry-java/blob/main/semconv/src/main/java/io/opentelemetry/semconv/trace/attributes/SemanticAttributes.java

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java [hunk 12]
```diff
@@ -256,7 +294,18 @@ protected void channelRead0(ChannelHandlerContext ctx, HttpRequest httpRequest)
...
+                span.setAttribute("HttpVersion",httpRequest.protocolVersion().protocolName());
+                span.setAttribute("RequestCode",requestCode);
```
Comment (tydhot) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java:308]: ues otel key instaed : https://github.com/open-telemetry/opentelemetry-java/blob/main/semconv/src/main/java/io/opentelemetry/semconv/trace/attributes/SemanticAttributes.java

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java [hunk 13]
```diff
@@ -298,7 +347,11 @@ protected void channelRead0(ChannelHandlerContext ctx, HttpRequest httpRequest)
                 processEventMeshRequest(ctx, asyncContext);
             } catch (Exception ex) {
                 httpServerLogger.error("AbrstractHTTPServer.HTTPHandler.channelRead0 err", ex);
-                // todo span end with exception.
+                //todo span end with exception.
+                span.addEvent("End Exceptional");
```
Comment (tydhot) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java:351]: we should add an exception here, but we need to record the cause in event atttribute

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/trace/OpenTelemetryTraceFactory.java [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    public OpenTelemetryTraceFactory(CommonConfiguration configuration){
+        try {
+            //different spanExporter
+            String exporterName = configuration.eventMeshTraceExporterType;
+            //use reflection to get spanExporter
+            String className = String.format("org.apache.eventmesh.runtime.exporter.%sExporter",exporterName);
+            EventMeshExporter eventMeshExporter = (EventMeshExporter) Class.forName(className).newInstance();
+            spanExporter = eventMeshExporter.getSpanExporter(configuration);
+        }catch (Exception ex){
+            logger.error("fail to set tracer's exporter,due to{}",ex.getMessage());
```
Comment (tydhot) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/trace/OpenTelemetryTraceFactory.java:64]: fail to set tracer's exporter,due to{} 
add a blank before{}

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java (hunk 7 above)
Comment (Roc-00) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java:145]: fixed

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java (hunk 8 above)
Comment (Roc-00) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java:157]: fixed

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java (hunk 9 above)
Comment (Roc-00) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java:161]: fixed

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java (hunk 10 above)
Comment (Roc-00) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java:300]: fixed

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java (hunk 11 above)
Comment (Roc-00) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java:306]: I have no idea why I can't import this class.
image
image

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/trace/OpenTelemetryTraceFactory.java (hunk 14 above)
Comment (Roc-00) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/trace/OpenTelemetryTraceFactory.java:64]: fixed

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java (hunk 13 above)
Comment (Roc-00) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java]: i am not sure whether i am doing it in a right way

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java (hunk 11 above)
Comment (Roc-00) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java]: fixed

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java (hunk 12 above)
Comment (Roc-00) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/AbstractHTTPServer.java]: fixed

File: eventmesh-common/src/test/resources/configuration.properties [hunk 15]
```diff
@@ -23,4 +23,6 @@ eventMesh.server.name=value5
 eventMesh.server.hostIp=value6
 eventMesh.connector.plugin.type=rocketmq
 eventMesh.security.plugin.type=acl
-eventMesh.registry.plugin.type=namesrv
\ No newline at end of file
+eventMesh.registry.plugin.type=namesrv
+eventMesh.exporter.zipkin.ip=localhost
```
Comment (qqeasonchen) [inline eventmesh-common/src/test/resources/configuration.properties:27]: keep the path the same as eventmesh.trace.export.zipkin.ip.

File: eventmesh-runtime/conf/eventmesh.properties [hunk 16]
```diff
@@ -71,4 +71,20 @@ eventMesh.registry.plugin.type=namesrv
...
+eventMesh.trace.exporter.type=Zipkin
+
+#set the maximum batch size to use
+eventMesh.trace.max.export.size=512
+#set the queue size. This must be >= the export batch size
+eventMesh.trace.max.queue.size=2048
+#set the max amount of time an export can run before getting(TimeUnit=SECONDS)
+eventMesh.trace.exporter.timeout=30
+#set time between two different exports(TimeUnit=SECONDS)
+eventMesh.trace.export.interval=5
```
Comment (qqeasonchen) [inline eventmesh-runtime/conf/eventmesh.properties:86]: keep the path the same as eventmesh.trace.exporter, eventMesh changed to eventmesh would be better.

File: eventmesh-common/src/test/resources/configuration.properties (hunk 15 above)
Comment (Roc-00) [reply eventmesh-common/src/test/resources/configuration.properties]: fixed

File: eventmesh-runtime/conf/eventmesh.properties (hunk 16 above)
Comment (Roc-00) [reply eventmesh-runtime/conf/eventmesh.properties]: fixed

**Review by xwm1992 — APPROVED** [review]

**Review by xwm1992 — APPROVED** [review]
LGTM

**Pull request merged**